In [1]:
import numpy as np
import pandas as pd
import mnist
import keras
from keras.datasets import fashion_mnist
from keras.models import Sequential
from keras.layers import Dense, Flatten
from keras.utils import to_categorical
from scikeras.wrappers import KerasClassifier
from sklearn.model_selection import GridSearchCV
from scikeras.wrappers import KerasClassifier

In [2]:
train_images = mnist.train_images()
train_labels = mnist.train_labels()
test_images = mnist.test_images()
test_labels = mnist.test_labels()

In [3]:
train_images = (train_images / 255)
test_images = (test_images / 255)

In [4]:
train_images = train_images.reshape((-1, 784))
test_images = test_images.reshape((-1, 784))

Создание модели

In [5]:
def create_model(optimizer='rmsprop'):
    model = Sequential([
        Dense(256, input_shape=(784,), activation='tanh'),
        Dense(128, activation='tanh'),
        Dense(64, activation='tanh'),
        Dense(32, activation='tanh'),
        Dense(10, activation='softmax')
    ])

    # Compile the model.
    model.compile(
        optimizer=optimizer,
        loss='categorical_crossentropy',
        metrics=['accuracy'],
    )
    return model


Обучение модели

In [6]:
model = create_model('sgd')
model.fit(train_images, to_categorical(train_labels), epochs=20, batch_size=64)

Epoch 1/20
938/938 [==============================] - 4s 3ms/step - loss: 0.7832 - accuracy: 0.8108
Epoch 2/20
938/938 [==============================] - 3s 3ms/step - loss: 0.3868 - accuracy: 0.8991
Epoch 3/20
938/938 [==============================] - 3s 3ms/step - loss: 0.3096 - accuracy: 0.9153
Epoch 4/20
938/938 [==============================] - 3s 3ms/step - loss: 0.2695 - accuracy: 0.9240
Epoch 5/20
938/938 [==============================] - 3s 3ms/step - loss: 0.2421 - accuracy: 0.9309
Epoch 6/20
938/938 [==============================] - 3s 3ms/step - loss: 0.2203 - accuracy: 0.9367
Epoch 7/20
938/938 [==============================] - 3s 3ms/step - loss: 0.2018 - accuracy: 0.9427
Epoch 8/20
938/938 [==============================] - 3s 3ms/step - loss: 0.1860 - accuracy: 0.9467
Epoch 9/20
938/938 [==============================] - 3s 3ms/step - loss: 0.1720 - accuracy: 0.9505
Epoch 10/20
938/938 [==============================] - 3s 3ms/step - loss: 0.1599 - accuracy: 0.9537

Оценка модели

In [7]:
loss, metrics = model.evaluate(test_images, to_categorical(test_labels))
print('**************')
print(f'Loss - {np.round(loss,2)}, Metrics - {np.round(metrics,3)}')

313/313 [==============================] - 1s 2ms/step - loss: 0.1051 - accuracy: 0.9680
**************
Loss - 0.11, Metrics - 0.968


Прогноз по первым 10 тестовым изображениям.

In [8]:
predictions = model.predict(test_images[:10])

1/1 [==============================] - 0s 97ms/step


In [9]:
print(f'Predictions: {np.argmax(predictions, axis=1)}')

Predictions: [7 2 1 0 4 1 4 9 5 9]


Сравнение моих прогнозов с реальными фактами

In [10]:
print(f'True labels: {test_labels[:10]}')

True labels: [7 2 1 0 4 1 4 9 5 9]


Изменения:
1. Я добавила 4 полносвязных слоя с 256, 128, 64 и 32 нейронами соответственно. Это позволит модели лучше обучиться на данных.
2. Вместо relu я использовала tanh в качестве функции активации. Это может помочь модели лучше обучиться на данных.
3. Вместо adam я использовала sgd (стохастический градиентный спуск) в качестве оптимизатора.
4. Я увеличила размер батча до 64 и количество эпох до 20, чтобы модель могла лучше обучиться на данных.

Результаты:
1. Точность на тестовом наборе: 0.968
2. Потери на тестовом наборе: 0.11, что является достаточно низким показателем. Это говорит о том, что модель хорошо обучилась и способна делать точные предсказания.
В целом, полученные результаты свидетельствуют о том, что модель показывает высокую производительность на данном наборе данных и может быть успешно использована для решения задачи классификации изображений.

Теперь попробуем с загруженной картинкой с цифрой.

In [35]:
import numpy as np
import matplotlib.pyplot as plt
from PIL import Image

# Загружаем изображение из папки
image = Image.open('number.png')

# Преобразуем изображение
image = image.convert('L')
image = image.resize((28, 28))
x = np.array(image)
x = x.reshape(1, 784)
x = x / 255.0

# Сделаем предсказание
prediction = model.predict(x)
predicted_class = np.argmax(prediction)

print(f"Модель предсказала, что это цифра: {predicted_class}")




1/1 [==============================] - 0s 24ms/step
Модель предсказала, что это цифра: 3
